producer/writer/message_writer.go

```diff
@@ -0,0 +1,378 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package writer
+
+import (
+	"container/list"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3msg/producer"
+	"github.com/m3db/m3x/clock"
+
+	"github.com/uber-go/tally"
+)
+
+const (
+	defaultAckMapSize = 1024
+)
+
+type messageWriter interface {
+	// Write writes the data.
+	Write(d producer.RefCountedData)
+
+	// Ack acknowledges the metadata.
+	Ack(meta metadata)
+
+	// Init initialize the message writer.
+	Init()
+
+	// Close closes the writer.
+	// It should block until all buffered data have been acknowledged.
+	Close()
+
+	// AddConsumerWriter adds a consumer writer for the given address.
+	AddConsumerWriter(addr string, cw consumerWriter)
+
+	// RemoveConsumerWriter removes the consumer writer for the given address.
+	RemoveConsumerWriter(addr string)
+
+	// ReplicatedShardID returns the replicated shard id.
+	ReplicatedShardID() uint64
+
+	// CutoverNanos returns the cutover nanoseconds.
+	CutoverNanos() int64
+
+	// SetCutoverNanos sets the cutover nanoseconds.
+	SetCutoverNanos(nanos int64)
+
+	// CutoffNanos returns the cutoff nanoseconds.
+	CutoffNanos() int64
+
+	// SetCutoffNanos sets the cutoff nanoseconds.
+	SetCutoffNanos(nanos int64)
+}
+
+type messageWriterMetrics struct {
+	consumerWriteError tally.Counter
+	writeError         tally.Counter
+	writeSuccess       tally.Counter
+	writeRetry         tally.Counter
+	writeNew           tally.Counter
+	writeAfterCutoff   tally.Counter
+	writeBeforeCutover tally.Counter
+}
+
+func newMessageWriterMetrics(scope tally.Scope) messageWriterMetrics {
+	return messageWriterMetrics{
+		consumerWriteError: scope.Counter("consumer-write-error"),
+		writeError:         scope.Counter("write-error"),
+		writeSuccess:       scope.Counter("write-success"),
+		writeRetry: scope.
+			Tagged(map[string]string{"write-type": "retry"}).
+			Counter("write"),
+		writeNew: scope.
+			Tagged(map[string]string{"write-type": "new"}).
+			Counter("write"),
+		writeAfterCutoff: scope.
+			Tagged(map[string]string{"reason": "after-cutoff"}).
+			Counter("invalid-write"),
+		writeBeforeCutover: scope.
+			Tagged(map[string]string{"reason": "before-cutover"}).
+			Counter("invalid-write"),
+	}
+}
+
+type messageWriterImpl struct {
+	sync.RWMutex
+
+	replicatedShardID uint64
+	mPool             messagePool
+	opts              Options
+
+	msgID           uint64
+	queue           *list.List
+	consumerWriters map[string]consumerWriter
+	acks            *acks
+	cutOffNanos     int64
+	cutOverNanos    int64
+	isClosed        bool
+	doneCh          chan struct{}
+	wg              sync.WaitGroup
+	m               messageWriterMetrics
+
+	nowFn clock.NowFn
+}
+
+func newMessageWriter(
+	replicatedShardID uint64,
```

Comment:
Is this the per-shard writer? Why is this called `replicatedShardID` if it can be used for the shared case as well?

Also we conventionally use `uint32` for shard ids.

Reply:
This is not per shard writer, the `shardWriter` is per shard, in a replicated mode, a `shardWriter` may own a few `messageWriter` which is differentiated by replicatedShardID, see pr description for examples for the replicated IDs.

Reply:
I see, perhaps just call `replicatedShardID` `replica`?

Reply:
I think `replicatedShardID` is more correct as `replica` should be 1,2,3 where `replicatedShardID` should be 123, 1123, 2123,